pallene/checker.lua

```diff
@@ -636,29 +638,47 @@ function FunChecker:check_var(var)
         elseif cname._tag == "checker.Name.Function" then
             var._type = self.p.module.functions[cname.id].typ
         elseif cname._tag == "checker.Name.Builtin" then
-            var._type = builtins[cname.name].typ
+            var._type = builtins.functions[cname.name].typ
         elseif cname._tag == "checker.Name.Module" then
             -- The type for a module is a table.
-            var._type = builtin_modules[cname.name]
+           var._type = builtins.modules[cname.name]
         else
-            error("Unexpected condition: cname._tag == " .. cname._tag)
+            error("impossible")
         end
 
+        -- print(inspect(var))
+
     elseif tag == "ast.Var.Dot" then
-        var.exp = self:check_exp_synthesize(var.exp)
-        local ind_type = var.exp._type
-        if not types.is_indexable(ind_type) then
-            type_error(var.loc,
-                "trying to access a member of value of type '%s'",
-                types.tostring(ind_type))
-        end
-        local field_type = types.indices(ind_type)[var.name]
-        if not field_type then
-            type_error(var.loc,
-                "field '%s' not found in type '%s'",
-                var.name, types.tostring(ind_type))
+        -- print(inspect(var))
+        -- print('---------------------')
+        if builtins.modules[var.exp.var.name] ~= nil then
+            local module_name = var.exp.var.name
+            local function_name = var.name
+            local internal_name = module_name .. "." .. function_name
+            local cname = self.p.symbol_table:find_symbol(internal_name)
+            return {
+                _name = cname,
+                _tag = "ast.Var.Name",
+                loc = var.exp.loc,
+                name = var.exp.var.name .. '.' .. var.name,
+                _type = builtins.functions[internal_name].typ
+            }
```

Comment:
We should use the `ast.Var.Name` constructor instead of building this table by hand:

     local flat_var = ast.Var.Name(var.exp.loc, internal_name)
     flat_var._name = ...
     flat_var._type = ...